### ISSUE TITLE
Ignore database columns not specified in the struct

### ISSUE THREAD

**Issue by multiplegeorges** [go-gorp/gorp#125]
I am working with some large legacy tables and only need a subset of the columns in my model.

It would be very useful to have gorp ignore columns that aren't mapped to a field in the struct.

Two ways of approaching this:

Change the default behaviour and just ignore unmapped columns. 
Create a config option on DbMap that enables allows incomplete models.

Thoughts?

**Comment by logan** [comment]
As is, it's unsafe to use "select *" if you ever plan on rolling out changes to the DB schema without downtime.

**Comment by fzerorubigd** [comment]
:-1: 
Personally, I prefer to use strict mapping, and prevent unpredictable behavior, like what you can see in language such as PHP.

**Comment by GeertJohan** [comment]
I agree. Required columns should be specified explicitly. What could be added though is a method to automatically query all required columns based on the struct.

**Comment by nelsam** [comment]
In case it helps, I've got a [project (which I've named gorq)](https://github.com/nelsam/gorq), which imports and extends gorp's functionality with a query DSL.  It was built to handle some of these sorts of issues (the whole project stems from a function I wrote to generate column names for use in queries).  It's new, kind of complicated, doesn't have the most useful error messages, needs some cleanup, and the test coverage isn't very solid yet; but it's being used actively on at least four projects (all of which are using postgresql).

The basic syntax of queries becomes (as an example):

```go
type Foo struct {
  ID int
  Bar Bar `db:"bar_id"` // Bar.ID populated with bar_id using sql.Scanner or a TypeConverter
  Baz string
}

foos := []Foo{}
ref := Foo{}
err := dbMap.Query(&ref).
  Where().
  Equal(&ref.Baz, "public").
  SelectToTarget(&foos)
```

It uses references quite a bit for figuring out column names (for input values, it will usually use the column name of a field if the value is equal to a field's address), which isn't exactly ideal... but I couldn't think of anything better at the time.  Still, it's a start.

More complicated queries (e.g. needing `OR` in the where clause, or using joins) are supported, but the syntax isn't always as clean as the above.  For example:

```go
foos = []Foo{}
err = dbMap.Query(&ref).
  Join(&ref.Bar).On().
  Equal(&ref.Bar, &ref.Bar.ID).
  Where().
  Greater(&ref.Bar.Rank, 3).
  SelectToTarget(&foos)

foos = []Foo{}
filter := filters.Or(filters.Equal(&ref.Bar, someAdminBarID), filters.Equal(&ref.Baz, "public"))
err = dbMap.Query(&ref).
  Where(filter).
  SelectToTarget(&foos)
```

I really wanted to have better test coverage before mentioning it to anyone else ... but I thought it might help anyone who's running in to this issue, and I can't hide forever.

By the way, my two cents, for what they're worth: `select *` is code smell, and discarding data (even via an option that defaults to off) is dangerous.  But my opinion's probably obvious from the fact that I wrote a pretty massive extension to gorp, in part to avoid using `select *`.

**Comment by GuyMorgan** [comment]
> I agree. Required columns should be specified explicitly. What could be added though is a method 
> to automatically query all required columns based on the struct.

Great, I want this and am willing to code it up. 

It needs to support the single/multiple row fetched and it should memoize the results of reflecting structs for keys. What's the API you'd like to see?

**Comment by GeertJohan** [comment]
Related to #133

**Comment by recht** [comment]
There is a very specific case where it would be nice if it was possible to ignore columns: when fetching database metadata where you don't have control of the output columns. For example, `show slave status` in mysql outputs a fairly long list of columns, and you can't limit the columns. You also don't quite know if a newer mysql version will add more columns to the output.

**Comment by nelsam** [comment]
@recht fetching metadata is done through `select` statements in postgres.  There are shortcuts in the `psql` CLI, but they all run select statements under the covers.

That said, if/when I find time to actually work on a new version of gorp in the future, I wouldn't mind adding this as a specific flag you can pass to queries.  Something along the lines of the following might make sense:

```
res, err := dbmap.Statement("show slave status")
if err != nil {
    // you know the drill
}
res.IgnoreMissing().Unmarshal(&slaveMeta)
```

Would that cover the edge case well enough for you?

**Comment by recht** [comment]
I know that postgresql uses regular selects for metadata. That's why I singled out mysql, which has custom statements for much of the metadata where you cannot control the output.

Anyway, your suggestion would definitely work for me.

**Comment by fiftin** [comment]
Hm, any updates for issue?
